p3387.tex

```diff
@@ -154,25 +160,25 @@ \subsection{Point of evaluation of \tcode{pre} and \tcode{post}}
 
-Note that in the non-coroutine wrapper \tcode{g}, this question cannot arise, because if a non-reference parameter of a user-defined non-coroutine function such as \tcode{g} is declared \tcode{const} on every declaration of \tcode{g}, the parameter object cannot be moved from or otherwise modified in the implementation of \tcode{g}; any an attempt to do so would be undefined behaviour as per \href{https://eel.is/c++draft/dcl.type.cv#4.sentence-1}{[dcl.type.cv]/4}.
+Note that in the non-coroutine wrapper \tcode{g}, this question cannot arise, because if a non-reference parameter of a user-defined non-coroutine function such as \tcode{g} is declared \tcode{const} on every declaration of \tcode{g}, the parameter object cannot be moved from or otherwise modified in the implementation of \tcode{g}; any attempt to do so would be undefined behaviour as per \href{https://eel.is/c++draft/dcl.type.cv#4.sentence-1}{[dcl.type.cv]/4}.
 %%%% OLD ABOVE
 
 
 In \cite{P2900R8}, in order to odr-use a non-reference parameter in the predicate of \tcode{post}, it has to be declared \tcode{const} on every declaration of the function. This allows us to reason about the parameter value not having been modified between the function call and the evaluation of that function's postcondition assertions, which is a prerequisite for writing meaningful postcondition assertions on such a parameter (see \cite{P2900R8}, Section 3.4.4). As it turns out, there is one implementation detail of coroutines that is unobservable caller-side today but leads to an interaction with the above rule if we wish to allow \tcode{post} on coroutines.
 
-When the ramp function is invoked, after initialising the parameters (and checking the precondition assertions, if we allow them), a copy of each parameter is created for the coroutine state, and the value of each parameter is moved into that copy (\href{https://eel.is/c++draft/dcl.fct.def.coroutine#13}{[dcl.fct.def.coroutine]/13}), potentially leaving the original parameter object in a moved-from state. This is one of the coroutine-specific operations that happen as part of the compiler-generated implementation of the ramp function, and is necessary to make the parameters accessible inside the coroutine body, whose lifetime can extend far beyond the ramp function returning.
+As part of the standard-mandated activity of the ramp, a copy of each parameter is created in the coroutine state, and the value of each parameter is moved into that copy (\href{https://eel.is/c++draft/dcl.fct.def.coroutine#13}{[dcl.fct.def.coroutine]/13}), potentially leaving the original parameter object in a moved-from state. This is necessary to make the parameters accessible inside the coroutine body, whose lifetime can extend far beyond the ramp function returning.
```

Comment:
we might want to mention that this is an elidable copy.  We can also just say "may be created in the coroutine stare" instead of "is created".

Reply:
In my first implementation, I made the parameter copies elidable (skipped when there were no body uses).  However, it was pointed out by Gor et. al. that the standard mandates the copy (one might argue that for some cases it's not observable whether the copy is made if it's never used).

It would be nice to have this optimisation - but the std does not appear to permit it.

Reply:
uh, class.copy.elision 1.3 is still a thing, right?

Reply:
If that's the case, let's leave the text as-is here.

Reply:
I'm happy to change it to "may be created", this is what we say in other places as well when we talk about these copies.

Reply:
> uh, class.copy.elision 1.3 is still a thing, right?

it is indeed - and mea culpa - things changed and I did not keep up.

Reply:
Of course, that paragraph (and the one that follows it) use the phrasing "if the meaning of the program would be unchanged..." which is... really awful wording.  It seems like even taking the address of those objects would change disalow copy elision in those cases.    That's certainly not the intent and the wording there should probably be made less wishy-washy and just allow copy elision without made-up conditions that are impossible to check.